File: backend/schemas/events.py
from pydantic import BaseModel
from typing import Optional, List, Dict, Any

class BaseEventSchema(BaseModel):
    type: str
    session_id: str
    timestamp: str

class SessionStartEventSchema(BaseEventSchema):
    type: str = "SESSION_START"
    topic: str
    prompt: str
    agents: List[Dict[str, Any]]
    config: Optional[Dict[str, Any]] = None

class ThinkStartEventSchema(BaseEventSchema):
    type: str = "THINK_START"
    agent_id: str

class ThinkEndEventSchema(BaseEventSchema):
    type: str = "THINK_END"
    agent_id: str

class TokenGrantedEventSchema(BaseEventSchema):
    type: str = "TOKEN_GRANTED"
    agent_id: str
    round_index: int
    turn_index: int

class ArgumentPostedEventSchema(BaseEventSchema):
    type: str = "ARGUMENT_POSTED"
    argument: Dict[str, Any]

class UpdateStartEventSchema(BaseEventSchema):
    type: str = "UPDATE_START"
    agent_id: str

class UpdateEndEventSchema(BaseEventSchema):
    type: str = "UPDATE_END"
    agent_id: str

class ThoughtUpdatedEventSchema(BaseEventSchema):
    type: str = "THOUGHT_UPDATED"
    thought: Dict[str, Any]

class TokenRequestEventSchema(BaseEventSchema):
    type: str = "TOKEN_REQUEST"
    agent_id: str
    novelty_tier: str
    priority_score: float
    position_in_queue: int

class QueueUpdatedEventSchema(BaseEventSchema):
    type: str = "QUEUE_UPDATED"
    queue: List[Dict[str, Any]]

class ConvergenceCheckEventSchema(BaseEventSchema):
    type: str = "CONVERGENCE_CHECK"
    status: str
    rounds_elapsed: int
    novel_claims_this_round: int

class SessionPausedEventSchema(BaseEventSchema):
    type: str = "SESSION_PAUSED"

class SessionResumedEventSchema(BaseEventSchema):
    type: str = "SESSION_RESUMED"

class SessionEndEventSchema(BaseEventSchema):
    type: str = "SESSION_END"
    reason: str
    rounds_elapsed: int
    summary_id: Optional[str] = None

class SummaryPostedEventSchema(BaseEventSchema):
    type: str = "SUMMARY_POSTED"
    summary: Dict[str, Any]

class ErrorEventSchema(BaseEventSchema):
    type: str = "ERROR"
    code: str
    message: str
    agent_id: Optional[str] = None

File: shared/types/events.ts
import { Agent, QueueEntry } from './agent';
import { SessionConfig } from './session';

export interface BaseEvent {
    type: string;
    session_id: string;
    timestamp: string;
}

export interface SessionStartEvent extends BaseEvent {
    type: 'SESSION_START';
    topic: string;
    prompt: string;
    agents: Agent[];
    config?: SessionConfig;
}

export interface ThinkStartEvent extends BaseEvent {
    type: 'THINK_START';
    agent_id: string;
}

export interface ThinkEndEvent extends BaseEvent {
    type: 'THINK_END';
    agent_id: string;
}

export interface TokenGrantedEvent extends BaseEvent {
    type: 'TOKEN_GRANTED';
    agent_id: string;
    round_index: number;
    turn_index: number;
}

export interface ArgumentPostedEvent extends BaseEvent {
    type: 'ARGUMENT_POSTED';
    argument: {
        id: string;
        agent_id: string;
        agent_name: string;
        round_index: number;
        turn_index: number;
        content: string;
    };
}

export interface UpdateStartEvent extends BaseEvent {
    type: 'UPDATE_START';
    agent_id: string;
}

export interface UpdateEndEvent extends BaseEvent {
    type: 'UPDATE_END';
    agent_id: string;
}

export interface ThoughtUpdatedEvent extends BaseEvent {
    type: 'THOUGHT_UPDATED';
    thought: {
        id: string;
        agent_id: string;
        version: number;
        content: string;
    };
}

export interface TokenRequestEvent extends BaseEvent {
    type: 'TOKEN_REQUEST';
    agent_id: string;
    novelty_tier: string;
    priority_score: number;
    position_in_queue: number;
}

export interface QueueUpdatedEvent extends BaseEvent {
    type: 'QUEUE_UPDATED';
    queue: QueueEntry[];
}

export interface ConvergenceCheckEvent extends BaseEvent {
    type: 'CONVERGENCE_CHECK';
    status: 'converging' | 'open' | 'capped';
    rounds_elapsed: number;
    novel_claims_this_round: number;
}

export interface SessionPausedEvent extends BaseEvent {
    type: 'SESSION_PAUSED';
}

export interface SessionResumedEvent extends BaseEvent {
    type: 'SESSION_RESUMED';
}

export interface SessionEndEvent extends BaseEvent {
    type: 'SESSION_END';
    reason: 'consensus' | 'cap' | 'host';
    rounds_elapsed: number;
    summary_id?: string;
}

export interface SummaryPostedEvent extends BaseEvent {
    type: 'SUMMARY_POSTED';
    summary: {
        id: string;
        content: string;
        termination_reason: 'consensus' | 'cap' | 'host';
    };
}

export interface ErrorEvent extends BaseEvent {
    type: 'ERROR';
    code: string;
    message: string;
    agent_id?: string;
}

export type RoundTableEvent =
    | SessionStartEvent
    | ThinkStartEvent
    | ThinkEndEvent
    | TokenGrantedEvent
    | ArgumentPostedEvent
    | UpdateStartEvent
    | UpdateEndEvent
    | ThoughtUpdatedEvent
    | TokenRequestEvent
    | QueueUpdatedEvent
    | ConvergenceCheckEvent
    | SessionPausedEvent
    | SessionResumedEvent
    | SessionEndEvent
    | SummaryPostedEvent
    | ErrorEvent;
